### ISSUE TITLE
Model's predictive performance depends upon system memory.

### ISSUE THREAD

**Issue by Nilabhra** [keras-team/keras#15531]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.6.0
- Python version: 3.7.12
- Bazel version (if compiling from source):
- GPU model and memory: Tesla P100-PCIE-16GB
- Exact command to reproduce:

**Describe the problem**.

Me and my colleague (@sayakpaul) are training the same composite neural network that we have written but we are observing major discrepancies in the train, validation and test performance (in terms of evaluation metrics). To ensure consistency, we are pulling the code from the same branch, using the same commit id and are not making any local changes. We are both using Google Colab Pro with a P100 GPU.

Digging deeper we realised that one of us was using a High RAM instance for the Colab notebook which resulted in giving much better results. The other person could replicate the result by switching to a High RAM instance. This was quite confusing as we were even using the same batch size.

Furthermore we investigated the effect of RAM and GPU and found that this behaviour is independent of GPU but very much dependant on RAM. A High RAM instance of Google Colab offers 25GB of RAM and apparently, the model converges nicely when the RAM is the same range (25-26 GB). We spun up a GCP Vertex AI Notebook with 30GB of RAM and tried training the model, but the performance was sub-par again. Then we tried the same with a ~26GB instance and the performance was back to being decent. Finally we reduced the RAM to 13 GB and got worse results as before.

The model comprises of the following components:
1. A pretrained BERT model from Tensorflow Hub (we are not updating the weights for this).
2. A GRU layer.
3. 2 IntegerLookup layers.
4. 2 Dense layers.

We read the data from TFRecords for training, validation and testing.

**Describe the current behavior**.

The neural network gives the best prediction performance (accuracy/precision) when trained in an instance with 25-26 GB of RAM.

**Describe the expected behavior**.

The size of RAM should not affect the validation and test performance of a model.

**Comment by sachinprasadhs** [comment]
@Nilabhra , any update with the code?

**Comment by Nilabhra** [comment]
@sachinprasadhs Hi! Unfortunately we didn't get the time to make the code available. It's going to be a bit of a hassle to do that.

We did realise that the memory might not have been the actual factor here, but rather the CPU cores are. It's a combination of CPU cores and the behaviour of tf.data.Dataset.interleave.

**Comment by tomerk** [comment]
Is your model's predictive performance sensitive to the ordering of the input datasets?

**Comment by Nilabhra** [comment]
> Is your model's predictive performance sensitive to the ordering of the input datasets?

We are realizing that it was the number of CPU cores rather than system memory that led to the discrepancy. Disabling interleaving of the tfrecords changes the performance drastically in any setting.

I'm closing this issue because the discrepancy is gone after using more data for training.
